File: packages/agent-mesh/sdks/typescript/tests/identity.test.ts
import { AgentIdentity } from '../src/identity';

describe('AgentIdentity', () => {
  let identity: AgentIdentity;

  beforeEach(() => {
    identity = AgentIdentity.generate('test-agent', ['read', 'write']);
  });

  describe('generate()', () => {
    it('creates an identity with a valid DID', () => {
      expect(identity.did).toMatch(/^did:agentmesh:test-agent:[0-9a-f]{16}$/);
    });

    it('creates an identity with a public key', () => {
      expect(identity.publicKey).toBeInstanceOf(Uint8Array);
      expect(identity.publicKey.length).toBeGreaterThan(0);
    });

    it('assigns capabilities', () => {
      expect(identity.capabilities).toEqual(['read', 'write']);
    });

    it('defaults to empty capabilities', () => {
      const id = AgentIdentity.generate('bare-agent');
      expect(id.capabilities).toEqual([]);
    });

    it('produces unique DIDs for different agents', () => {
      const other = AgentIdentity.generate('other-agent');
      expect(other.did).not.toEqual(identity.did);
    });
  });

  describe('sign() / verify()', () => {
    it('signs and verifies data correctly', () => {
      const data = new TextEncoder().encode('hello agentmesh');
      const signature = identity.sign(data);

      expect(signature).toBeInstanceOf(Uint8Array);
      expect(signature.length).toBeGreaterThan(0);
      expect(identity.verify(data, signature)).toBe(true);
    });

    it('rejects tampered data', () => {
      const data = new TextEncoder().encode('original');
      const signature = identity.sign(data);
      const tampered = new TextEncoder().encode('tampered');

      expect(identity.verify(tampered, signature)).toBe(false);
    });

    it('rejects invalid signature', () => {
      const data = new TextEncoder().encode('test');
      const badSig = new Uint8Array(64).fill(0);

      expect(identity.verify(data, badSig)).toBe(false);
    });
  });

  describe('toJSON() / fromJSON()', () => {
    it('round-trips through JSON serialization', () => {
      const json = identity.toJSON();
      const restored = AgentIdentity.fromJSON(json);

      expect(restored.did).toEqual(identity.did);
      expect(restored.capabilities).toEqual(identity.capabilities);
    });

    it('produces valid JSON fields', () => {
      const json = identity.toJSON();
      expect(json).toHaveProperty('did');
      expect(json).toHaveProperty('publicKey');
      expect(json).toHaveProperty('privateKey');
      expect(json).toHaveProperty('capabilities');
      expect(typeof json.publicKey).toBe('string');
    });

    it('restored identity can sign and verify', () => {
      const json = identity.toJSON();
      const restored = AgentIdentity.fromJSON(json);
      const data = new TextEncoder().encode('roundtrip test');
      const sig = restored.sign(data);

      expect(restored.verify(data, sig)).toBe(true);
    });
  });
});

File: packages/agent-mesh/sdks/go/identity_test.go
package agentmesh

import (
	"testing"
)

func TestGenerateIdentity(t *testing.T) {
	id, err := GenerateIdentity("agent-1", []string{"read", "write"})
	if err != nil {
		t.Fatalf("GenerateIdentity: %v", err)
	}
	if id.DID != "did:agentmesh:agent-1" {
		t.Errorf("DID = %q, want did:agentmesh:agent-1", id.DID)
	}
	if len(id.PublicKey) != 32 {
		t.Errorf("PublicKey length = %d, want 32", len(id.PublicKey))
	}
	if len(id.Capabilities) != 2 {
		t.Errorf("Capabilities length = %d, want 2", len(id.Capabilities))
	}
}

func TestSignAndVerify(t *testing.T) {
	id, err := GenerateIdentity("signer", nil)
	if err != nil {
		t.Fatal(err)
	}

	data := []byte("hello agent mesh")
	sig, err := id.Sign(data)
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	if !id.Verify(data, sig) {
		t.Error("Verify returned false for valid signature")
	}
	if id.Verify([]byte("tampered"), sig) {
		t.Error("Verify returned true for tampered data")
	}
}

func TestSignWithoutPrivateKey(t *testing.T) {
	id := &AgentIdentity{DID: "did:agentmesh:nopk"}
	_, err := id.Sign([]byte("data"))
	if err == nil {
		t.Error("expected error when signing without private key")
	}
}

func TestJSONRoundTrip(t *testing.T) {
	id, _ := GenerateIdentity("json-agent", []string{"cap1"})
	data, err := id.ToJSON()
	if err != nil {
		t.Fatalf("ToJSON: %v", err)
	}

	restored, err := FromJSON(data)
	if err != nil {
		t.Fatalf("FromJSON: %v", err)
	}
	if restored.DID != id.DID {
		t.Errorf("DID mismatch: %q vs %q", restored.DID, id.DID)
	}
	if len(restored.PublicKey) != len(id.PublicKey) {
		t.Error("PublicKey length mismatch after round-trip")
	}
}
